runtime/databricks/automl_runtime/forecast/pmdarima/training.py

```diff
@@ -130,8 +131,30 @@ def _fit_predict(df: pd.DataFrame, cutoffs: List[pd.Timestamp], seasonal_period:
 
     @staticmethod
     def _fill_missing_time_steps(df: pd.DataFrame, frequency: str):
+        df_filled = df.set_index("ds")
+        # NOTE: We have to normalize the index before resampling since pandas will
+        # normalize the week month quarter and year data during resample causing
+        # the closed right not work as expected.
+        # TODO(ML-27585): Improve this implementation/file a PR to pandas.
+        if OFFSET_ALIAS_MAP[frequency] in ['W', 'MS', 'QS', 'YS']:
+            df_filled.index = df_filled.index.normalize()
         # Forward fill missing time steps
-        df_filled = df.set_index("ds").resample(rule=OFFSET_ALIAS_MAP[frequency]).pad().reset_index()
+        # NOTE: the right closed meanning that the time data after resample
+        # will be right-shifted to the closest frequency, e.g. 2020-01-01 03:00:00
+        # will be shifted to 2020-01-02 00:00:00, thus we can perform a forward
+        # fill to fill the NaN since all the resampled time are after the original
+        # time. 
+        # Reference: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.resample.html
+        df_filled = df_filled.resample(
+                rule=OFFSET_ALIAS_MAP[frequency], closed='right'
+            ).pad().reset_index()
+        # We want to re-align the resampled time to the original start time.
+        # However, since all of our supported monthly/quarterly/annualy data
+        # are counted from the begining of month/quarter/year, so we need to
+        # first shift back one frequency unit to correctly caluclate the offset. 
+        df_filled["ds"] = df_filled["ds"] - pd.DateOffset(
+            **DATE_OFFSET_KEYWORD_MAP[OFFSET_ALIAS_MAP[frequency]]
+        )
```

Comment:
Could you give an example why we need to do this shift? I don't get it completely.

Reply:
Done.